### PR TITLE
expression, plan: restore WrapCastAsReal and WrapCastAsDecimal when arg is ClassInt

### PR DESCRIPTION
**Pull request by XuHuaiyu**
The PR #4200 sets `flen` to the flen of the arg in WrapWithCastAsDecimal and WrapWithCastAsReal
when arg is `ClassInt`,
**BUT**,
when the arg is a scalar function which returns an int result **with a wrong flen**,
may cause an unexpected truncate and produce a wrong eval result.

So I restore the changes and add a test case in the commit.

PTAL @jackysp @zz-jason @hanfei1991

### CODE REVIEW

**Comment by shenli** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
@shenli 
Yeah, this commit tries to fix sql logical test

mysql-test should be fixed in tidb-test

**Comment by hanfei1991** [comment]
LGTM

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1294,9 +1294,6 @@ func WrapWithCastAsReal(expr Expression, ctx context.Context) (Expression, error
 	}
 	tp := types.NewFieldType(mysql.TypeDouble)
 	tp.Flen, tp.Decimal = mysql.MaxRealWidth, types.UnspecifiedLength
-	if exprTp := expr.GetType(); exprTp.ToClass() == types.ClassInt {
```
Comment (zz-jason) [inline expression/builtin_cast.go]: how about add a `TODO` to indicate we should revert this change after all functions being rewrited ?
